<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Helper Tool &mdash; MacSyFinder 2.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/_css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Publishing/sharing models" href="publish_package.html" />
    <link rel="prev" title="Macromolecular models" href="modeling.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> MacSyFinder
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../user_guide/index.html">User Guide</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Modeller Guide</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html#modelling-systems-with-macsyfinder">Modelling Systems with MacSyFinder</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l3"><a class="reference internal" href="package.html">Models Package</a></li>
<li class="toctree-l3"><a class="reference internal" href="modeling.html">Macromolecular models</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Helper Tool</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#macsyprofile">macsyprofile</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="publish_package.html">Publishing/sharing models</a></li>
<li class="toctree-l3"><a class="reference internal" href="gallery.html">Gallery of examples of MacSyFinder’s models</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#carrying-models-from-v1-to-v2">Carrying models from v1 to v2</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#frequently-asked-questions">Frequently Asked Questions</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/index.html">Developer Guide</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MacSyFinder</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Modeller Guide</a></li>
      <li class="breadcrumb-item active">Helper Tool</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/modeler_guide/helper_tools.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="helper-tool">
<span id="id1"></span><h1>Helper Tool<a class="headerlink" href="#helper-tool" title="Permalink to this heading"></a></h1>
<section id="macsyprofile">
<span id="id2"></span><h2>macsyprofile<a class="headerlink" href="#macsyprofile" title="Permalink to this heading"></a></h2>
<p>To help develop new models we provide the tool <cite>macsyprofile</cite> which is to be used as post treatement.</p>
<p>It is ran over a previous macsyfinder analysis:</p>
<blockquote>
<div><ul class="simple">
<li><p>it extracts from raw HMMER output files the hits and computes the profile coverage for each of them.</p></li>
<li><p>it enables to filter the hits in a user-defined manner, to test other values of filtering parameters than those used with the MacSyFinder run.</p></li>
<li><p>it writes down the results in a file in <cite>tsv</cite> format <cite>hmm_coverage.tsv</cite>.</p></li>
</ul>
</div></blockquote>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>usage: macsyprofile [-h] [--coverage-profile COVERAGE_PROFILE]
                    [--i-evalue-sel I_EVALUE_SEL]
                    [--best-hits {score,i_eval,profile_coverage}] [-p PATTERN]
                    [-o OUT] [-f] [-V] [-v] [--mute]
                    previous_run

     *            *               *                   * *       *
*           *               *   *   *  *    **                *
  **     *    *   *  *     *                    *               *
     __  __    *       ____ *      ____        **   __ _ _  *
    |  \/  | __ _  ___/ ___| _   _|  _ \ _ __ ___  / _(_) | ___
    | |\/| |/ _` |/ __\___ \| | | | |_) | &#39;__/ _ \| |_| | |/ _ \
    | |  | | (_| | (__ ___) | |_| |  __/| | | (_) |  _| | |  __/
    |_|  |_|\__,_|\___|____/ \__, |_|   |_|  \___/|_| |_|_|\___|
            *                |___/    *                   *
 *      *   * *     *   **         *   *  *           *
  *      *         *        *    *              *
             *                           *  *           *     *

MacSyProfile - MacSyFinder profile helper tool

positional arguments:
  previous_run          The path to a macsyfinder results directory.

optional arguments:
  -h, --help            show this help message and exit
  --coverage-profile COVERAGE_PROFILE
                        Minimal profile coverage required for the hit
                        alignment with the profile to allow the hit selection
                        for systems detection. (default no threshold)
  --i-evalue-sel I_EVALUE_SEL
                        Maximal independent e-value for Hmmer hits to be
                        selected for systems detection. (default: no selection
                        based on i-evalue)
  --best-hits {score,i_eval,profile_coverage}
                        If several hits match the same replicon, same gene.
                        Select only the best one (based on best &#39;score&#39; or
                        &#39;i_evalue&#39; or &#39;profile_coverage&#39;)
  -p PATTERN, --pattern PATTERN
                        pattern to filter the hmm files to analyse.
  -o OUT, --out OUT     the path to a file to write results.
  --index-dir INDEX_DIR
                        Specifies the path to a directory to store/read the sequence index
                        when the sequence-db dir is not writable.
  -f, --force           force to write output even the file already exists
                        (overwrite it).
  -V, --version         show program&#39;s version number and exit
  -v, --verbosity       Increases the verbosity level. There are 4 levels:
                        Error messages (default), Warning (-v), Info (-vv) and
                        Debug.(-vvv)
  --mute                Mute the log on stdout. (continue to log on
                        macsyfinder.log) (default: False)

For more details, visit the MacSyFinder website and see the MacSyFinder documentation.
</pre></div>
</div>
<p>For instance:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>&gt;macsyprofile  macsyfinder-2021XXXX_XX-XX-XX
</pre></div>
</div>
<p>will analyse the HMMER raw outputs stored in <cite>macsyfinder-2021XXXX_XX-XX-XX/hmmer_results</cite> directory
and the results will be stored in <cite>macsyfinder-2021XXXX_XX-XX-XX/hmm_coverage.tsv</cite> file</p>
<section id="setting-filtering-parameters">
<h3>Setting filtering parameters<a class="headerlink" href="#setting-filtering-parameters" title="Permalink to this heading"></a></h3>
<p>This helper tool is designed to help the user test the relevance of the HMM profiles used, what filtering parameters for HMMER to be used, and understand why some components might be unexpectedly missing from the MacSyFinder results.
This can thus help to improve the models - for instance for the genomic location parameters (is a component not found cause it should be listed as a <cite>loner</cite>?).</p>
<p>Therefore by default, the filtering parameters are very loose so that most hits found with HMMER will be reported, even the weakest ones.</p>
<p>However, it is possible to filter hits to be extracted based on the profile coverage with <cite>–coverage-profile</cite> or the i-evalue (<cite>–i-evalue-sel</cite>) to be a bit more stringent.</p>
<p>Also, it is possible to use the <cite>–best-hits</cite> in order to report only the best hit for a given protein sequence when several profiles were matching hit.</p>
</section>
<section id="using-patterns-with-pattern">
<h3>Using patterns with “–pattern”<a class="headerlink" href="#using-patterns-with-pattern" title="Permalink to this heading"></a></h3>
<p>If in <cite>previous_run/hmmer_results</cite> you have the following files:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>previous_run/hmmer_results/Archaeal-T4P_arCOG11238.search_hmm.out
previous_run/hmmer_results/Archaeal-T4P_arCOG11520.search_hmm.out
previous_run/hmmer_results/Archaeal-T4P_arCOG11777.search_hmm.out
previous_run/hmmer_results/Archaeal-T4P_arCOG11778.search_hmm.out
previous_run/hmmer_results/Archaeal-T4P_arCOG11936.search_hmm.out
previous_run/hmmer_results/Archaeal-T4P_arCOG14515.search_hmm.out
previous_run/hmmer_results/ComM_comC.search_hmm.out
previous_run/hmmer_results/ComM_comEB.search_hmm.out
previous_run/hmmer_results/ComM_comEC.search_hmm.out
previous_run/hmmer_results/ComM_comGA.search_hmm.out
previous_run/hmmer_results/ComM_comGB.search_hmm.out
previous_run/hmmer_results/ComM_comGC.search_hmm.out
previous_run/hmmer_results/ComM_comGD.search_hmm.out
previous_run/hmmer_results/ComM_comGE.search_hmm.out
previous_run/hmmer_results/MSH_mshA.search_hmm.out
previous_run/hmmer_results/MSH_mshB.search_hmm.out
previous_run/hmmer_results/MSH_mshC.search_hmm.out
</pre></div>
</div>
<p>But you are interested only in ComM family genes, you can specify the option <code class="docutils literal notranslate"><span class="pre">--pattern</span> <span class="pre">'ComM*'</span></code>
For instance:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&gt;macsyprofile --pattern &#39;ComM*&#39;  macsyfinder-2021XXXX_XX-XX-XX
parsing macsyfinder-2021XXXX_XX-XX-XX/hmmer_results/ComM_comB.search_hmm.out
parsing macsyfinder-2021XXXX_XX-XX-XX/hmmer_results/ComM_comC.search_hmm.out
parsing macsyfinder-2021XXXX_XX-XX-XX/hmmer_results/ComM_comEA.search_hmm.out
parsing macsyfinder-2021XXXX_XX-XX-XX/hmmer_results/ComM_comEB.search_hmm.out
parsing macsyfinder-2021XXXX_XX-XX-XX/hmmer_results/ComM_comEC.search_hmm.out
parsing macsyfinder-2021XXXX_XX-XX-XX/hmmer_results/ComM_comGA.search_hmm.out
parsing macsyfinder-2021XXXX_XX-XX-XX/hmmer_results/ComM_comGB.search_hmm.out
parsing macsyfinder-2021XXXX_XX-XX-XX/hmmer_results/ComM_comGC.search_hmm.out
parsing macsyfinder-2021XXXX_XX-XX-XX/hmmer_results/ComM_comGD.search_hmm.out
parsing macsyfinder-2021XXXX_XX-XX-XX/hmmer_results/ComM_comGE.search_hmm.out
found 79 hits
result is in &#39;macsyfinder-2021XXXX_XX-XX-XX/hmm_coverage.tsv&#39;
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The patterns available are the <cite>glob</cite> patterns (the jokers usable with unix <cite>ls</cite> command )</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&gt;macsyprofile --pattern &#39;ComM_com?C&#39; -f macsyfinder-2021XXXX_XX-XX-XX
parsing macsyfinder-2021XXXX_XX-XX-XX/hmmer_results/ComM_comEC.search_hmm.out
parsing macsyfinder-2021XXXX_XX-XX-XX/hmmer_results/ComM_comGC.search_hmm.out
found 16 hits
result is in &#39;macsyfinder-2021XXXX_XX-XX-XX/hmm_coverage.tsv&#39;
</pre></div>
</div>
</div>
</section>
<section id="a-useful-example-for-modellers">
<h3>A useful example for modellers?<a class="headerlink" href="#a-useful-example-for-modellers" title="Permalink to this heading"></a></h3>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&gt;macsyprofile --best-hits i_eval --i-evalue-sel 0.001 --coverage-profile 0.5 -o msf_GCF_003149495.1_ASM314949v1_tff-sf/hmm_coverage_best-hits_ieval_default_filter_MSF.tsv msf_GCF_003149495.1_ASM314949v1_tff-sf
found 221 hits
result is in &#39;msf_GCF_003149495.1_ASM314949v1_tff-sf/hmm_coverage_best-hits_ieval_default_filter_MSF.tsv&#39;
</pre></div>
</div>
<p>This command line might be useful to macsy-models modellers, as it consists in extracting all relevant hits that are used by
the MacSyFinder engine to search systems, when using the default parameters:</p>
<ul class="simple">
<li><p>the proteins are assigned with their best hits (i-evalue based) when they match several profiles (<cite>–best-hits i_eval</cite> option)</p></li>
<li><p>the default filtering parameters (i-evalue and profile coverage) are used (<cite>–i-evalue-sel</cite> and <cite>–coverage-profile</cite> options)</p></li>
</ul>
<p>By using this command line that lists all hits available for MacSyFinder to search for systems, one could be interested in
comparing this list to the list of hits that end in being assigned to systems (listed e.g. in best_solution.tsv).
This can help to determine why a component is missing from a system: is it because there are no good hits for it, or
is it because it does not comply to the co-localization rules defined in the systems’ model?</p>
</section>
<section id="parsing-macsyprofile-outputs">
<h3>Parsing macsyprofile outputs<a class="headerlink" href="#parsing-macsyprofile-outputs" title="Permalink to this heading"></a></h3>
<p>The <cite>macsyprofile</cite> output is a tabulated separated values (<cite>.tsv</cite>) files
The first lines which are comments (starting with ‘#’) display the tool version
and the complete command line used. Then follow the results.
The first line of results is a header line.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span># macsyprofile 2.0rc1
# macsyprofile --pattern ComM* --coverage-profile 0.5 macsyfinder-20201202_15-17-46/
hit_id  replicon_name   position_hit    hit_sequence_length     gene_name       i_eval  score   profile_coverage        sequence_coverage       begin   end
GCF_000006745_021980    GCF_000006745   2198    291     ComM_comC       2.500e-40       136.400 0.942   0.708   62      267
GCF_000006745_007650    GCF_000006745   765     253     ComM_comC       9.600e-31       105.100 0.937   0.798   43      244
...
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This file can be easily parsed using the Python <a class="reference external" href="https://pandas.pydata.org/">pandas</a> library.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">systems</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;path/to/hmm_coverage.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The <cite>macsyprofile</cite> tool is not compliant with results produced with <cite>macsyfinder v1</cite>.
If you get <code class="docutils literal notranslate"><span class="pre">Cannot</span> <span class="pre">find</span> <span class="pre">models</span> <span class="pre">in</span> <span class="pre">conf</span> <span class="pre">file</span> <span class="pre">XXX.</span> <span class="pre">May</span> <span class="pre">be</span> <span class="pre">these</span> <span class="pre">results</span> <span class="pre">have</span> <span class="pre">been</span> <span class="pre">generated</span> <span class="pre">with</span> <span class="pre">an</span> <span class="pre">old</span> <span class="pre">version</span> <span class="pre">of</span> <span class="pre">macsyfinder.</span></code>
Check the configuration file, if <cite>[models]</cite> section contains <code class="docutils literal notranslate"><span class="pre">models_1</span> <span class="pre">=</span> <span class="pre">XXX</span> <span class="pre">YYY</span></code> remove the <cite>_1</cite> from models
<code class="docutils literal notranslate"><span class="pre">models</span> <span class="pre">=</span> <span class="pre">XXX</span> <span class="pre">YYY</span></code></p>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="modeling.html" class="btn btn-neutral float-left" title="Macromolecular models" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="publish_package.html" class="btn btn-neutral float-right" title="Publishing/sharing models" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2014-2024, &#39;Institut Pasteur (Paris), CNRS&#39;.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
  

<div class="licensing">
  <p>
    This documentation is licensed under <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License
      <img src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png">.
    </a>
  </p>
</div>


</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

<a href="https://github.com/gem-pasteur/macsyfinder">
    <img style="position: absolute; top: 0; right: 0; border: 0;"
         src="https://camo.githubusercontent.com/38ef81f8aca64bb9a64448d0d70f1308ef5341ab/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f72696768745f6461726b626c75655f3132313632312e706e67"
         alt="Fork me on GitHub"
         data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png">
</a>



</body>
</html>