// dpmol/lib/mol-plugin-state/manager/structure/grid-pocket.js
"use strict";
/**
 * Copyright (c) 2019-2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GridPocketOrderLabelTag = exports.GridPocketGroupTag = exports.StructureGridPocketManager = void 0;
var tslib_1 = require("tslib");
var mol_state_1 = require("../../../mol-state");
var transforms_1 = require("../../transforms");
var commands_1 = require("../../../mol-plugin/commands");
var component_1 = require("../../component");
exports.GridPocketGroupTag = 'grid-pocket-group';
exports.GridPocketOrderLabelTag = 'grid-pocket-order-label';
var StructureGridPocketManager = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(StructureGridPocketManager, _super);
    function StructureGridPocketManager(plugin) {
        var _this = _super.call(this, {
            boxes: [],
        }) || this;
        _this.plugin = plugin;
        _this.behaviors = {
            state: _this.ev.behavior(_this.state)
        };
        _this._empty = [];
        plugin.state.data.events.changed.subscribe(function (e) {
            if (e.inTransaction || plugin.behaviors.state.isAnimating.value)
                return;
            _this.sync();
        });
        plugin.behaviors.state.isAnimating.subscribe(function (isAnimating) {
            if (!isAnimating && !plugin.behaviors.state.isUpdating.value)
                _this.sync();
        });
        return _this;
    }
    StructureGridPocketManager.prototype.stateUpdated = function () {
        this.behaviors.state.next(this.state);
    };
    StructureGridPocketManager.prototype.getGroup = function () {
        var state = this.plugin.state.data;
        var groupRef = mol_state_1.StateSelection.findTagInSubtree(state.tree, mol_state_1.StateTransform.RootRef, exports.GridPocketGroupTag);
        var builder = this.plugin.state.data.build();
        if (groupRef)
            return builder.to(groupRef);
        return builder
            .toRoot()
            .group(transforms_1.StateTransforms.Misc.CreateGroup, { label: "GridPocket" }, { tags: exports.GridPocketGroupTag });
    };
    StructureGridPocketManager.prototype.addBox = function (center, x, y, z, cRef, options) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var update, res, state;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        update = cRef ? this.plugin.state.data.build().to(cRef) : this.getGroup();
                        return [4 /*yield*/, update
                                .apply(transforms_1.StateTransforms.Model.MultiStructureSelectionFromExpression, { label: 'GridPocketStructure' })
                                .apply(transforms_1.StateTransforms.Representation.StructureGridPocket3D, { x: x, y: y, z: z, center: center, })];
                    case 1:
                        res = _a.sent();
                        state = this.plugin.state.data;
                        return [4 /*yield*/, commands_1.PluginCommands.State.Update(this.plugin, {
                                state: state,
                                tree: update,
                                options: { doNotLogTiming: true }
                            })];
                    case 2:
                        _a.sent();
                        return [2 /*return*/, res];
                }
            });
        });
    };
    StructureGridPocketManager.prototype.getTransforms = function (transformer) {
        var state = this.plugin.state.data;
        var groupRef = mol_state_1.StateSelection.findTagInSubtree(state.tree, mol_state_1.StateTransform.RootRef, exports.GridPocketGroupTag);
        var ret = groupRef
            ? state.select(mol_state_1.StateSelection.Generators.ofTransformer(transformer, groupRef))
            : this._empty;
        if (ret.length === 0)
            return this._empty;
        return ret;
    };
    StructureGridPocketManager.prototype.sync = function () {
        var labels = [];
        for (var _i = 0, _a = this.getTransforms(transforms_1.StateTransforms.Representation.StructureSelectionsLabel3D); _i < _a.length; _i++) {
            var cell = _a[_i];
            var tags = cell.obj['tags'];
            if (!tags || !tags.includes(exports.GridPocketOrderLabelTag))
                labels.push(cell);
        }
        var updated = this.updateState({
            boxes: this.getTransforms(transforms_1.StateTransforms.Representation.StructureGridPocket3D),
        });
        if (updated)
            this.stateUpdated();
    };
    return StructureGridPocketManager;
}(component_1.StatefulPluginComponent));
exports.StructureGridPocketManager = StructureGridPocketManager;
