// dpmol/lib/light-plugin/representation/molecular/ball-stick.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BallAndStickRepr = void 0;
var tslib_1 = require("tslib");
var ball_and_stick_1 = require("../../../mol-repr/structure/representation/ball-and-stick");
exports.BallAndStickRepr = {
    create: function (structure, ctx, theme) {
        var _a;
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var representation;
            return (0, tslib_1.__generator)(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        representation = ball_and_stick_1.BallAndStickRepresentationProvider.factory(ctx, ball_and_stick_1.BallAndStickRepresentationProvider.getParams);
                        representation.setTheme(theme !== null && theme !== void 0 ? theme : {
                            color: ctx.colorThemeRegistry.create('element-symbol', {
                                structure: structure
                            }),
                            size: ctx.sizeThemeRegistry.create('uniform', {
                                structure: structure
                            })
                        });
                        return [4 /*yield*/, representation
                                .createOrUpdate((0, tslib_1.__assign)((0, tslib_1.__assign)({}, ball_and_stick_1.BallAndStickRepresentationProvider.defaultValues), { aromaticBonds: ((_a = structure.model) === null || _a === void 0 ? void 0 : _a.sourceData.kind) === 'mmCIF', quality: 'auto' }), structure)
                                .run()];
                    case 1:
                        _b.sent();
                        return [2 /*return*/, representation];
                }
            });
        });
    },
    update: function (representation, structure, ctx, theme) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        representation.setTheme(theme !== null && theme !== void 0 ? theme : {
                            color: ctx.colorThemeRegistry.create('element-symbol', {
                                structure: structure
                            }),
                            size: ctx.sizeThemeRegistry.create('uniform', {
                                structure: structure
                            })
                        });
                        return [4 /*yield*/, representation.createOrUpdate({}, structure).run()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    }
};
