// dpmol/lib/light-plugin/representation/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.INTERACTION_TYPE = exports.OtherRepresentation = exports.MeasurementRepresentation = exports.MolecularRepresentation = void 0;
var types_1 = require("../types");
var ribbon_1 = require("./molecular/ribbon");
var ball_stick_1 = require("./molecular/ball-stick");
var label_1 = require("./measurement/label");
var distance_1 = require("./measurement/distance");
var angle_1 = require("./measurement/angle");
var dihedral_1 = require("./measurement/dihedral");
var line_1 = require("./molecular/line");
var stick_1 = require("./molecular/stick");
var surface_1 = require("./surface");
var rotate_arrow_1 = require("./shape/rotate-arrow");
var cpk_1 = require("./molecular/cpk");
var pocket_1 = require("./shape/pocket");
var pharmacophore_1 = require("./shape/pharmacophore");
var iso_surface_1 = require("./shape/iso-surface");
var mapping_line_1 = require("./shape/mapping-line");
exports.MolecularRepresentation = (_a = {},
    _a[types_1.RepresentationType.Ribbon] = ribbon_1.RibbonRepr,
    _a[types_1.RepresentationType.BallAndStick] = ball_stick_1.BallAndStickRepr,
    _a[types_1.RepresentationType.Line] = line_1.LineRepr,
    _a[types_1.RepresentationType.Stick] = stick_1.StickRepr,
    _a[types_1.RepresentationType.CPK] = cpk_1.CPKRepr,
    _a);
exports.MeasurementRepresentation = (_b = {},
    _b[types_1.RepresentationType.Label] = label_1.LabelRepr,
    _b[types_1.RepresentationType.Distance] = distance_1.DistanceRepr,
    _b[types_1.RepresentationType.Angle] = angle_1.AngleRepr,
    _b[types_1.RepresentationType.Dihedral] = dihedral_1.DihedralRepr,
    _b);
exports.OtherRepresentation = (_c = {},
    _c[types_1.RepresentationType.Surface] = surface_1.SurfaceRepr,
    _c[types_1.RepresentationType.IsoSurface] = iso_surface_1.IsoSurfaceRepr,
    _c[types_1.RepresentationType.RotateArrow] = rotate_arrow_1.RotateArrowRepr,
    _c[types_1.RepresentationType.Pocket] = pocket_1.PocketRepr,
    _c[types_1.RepresentationType.Pharmacophore] = pharmacophore_1.PharmachophoreRepr,
    _c[types_1.RepresentationType.MappingLine] = mapping_line_1.MappingLineRepr,
    _c);
var INTERACTION_TYPE;
(function (INTERACTION_TYPE) {
    INTERACTION_TYPE["LigandReceptor"] = "ligand-receptor";
    INTERACTION_TYPE["IntraLigand"] = "intra-ligand";
    INTERACTION_TYPE["IntraReceptor"] = "intra-receptor";
    INTERACTION_TYPE["All"] = "all";
})(INTERACTION_TYPE = exports.INTERACTION_TYPE || (exports.INTERACTION_TYPE = {}));
